gpuallocator/besteffort_policy.go

```diff
@@ -29,33 +31,55 @@ func NewBestEffortPolicy() Policy {
 //  Such a solution is necessary in the general case because of the
 //  non-hierarchical nature of the various links that influence the score
 //  calculated for each pair of GPUs.
-func (p *bestEffortPolicy) Allocate(devices []*Device, size int) []*Device {
+func (p *bestEffortPolicy) Allocate(available []*Device, required []*Device, size int) []*Device {
 	if size <= 0 {
 		return []*Device{}
 	}
 
-	if len(devices) < size {
+	if len(available) < size {
+		return []*Device{}
+	}
+
+	if len(required) > size {
 		return []*Device{}
 	}
 
 	// Find the highest scoring GPU partition with sets of of size 'size'.
+	// Don't consider partitions that don't have at least one set that contains
+	// all of the GPUs 'required' by the allocation.
 	bestPartition := [][]*Device(nil)
 	bestScore := 0
-	iterateGPUPartitions(devices, size, func(candidate [][]*Device) {
+	iterateGPUPartitions(available, size, func(candidate [][]*Device) {
+		if !gpuPartitionContainsSetWithAll(candidate, required) {
+			return
+		}
 		score := calculateGPUPartitionScore(candidate)
 		if score > bestScore || bestPartition == nil {
 			bestPartition = candidate
 			bestScore = score
 		}
 	})
 
+	// Filter the 'bestPartition' to only include sets containing all of the
+	// 'required' devices (which may be nil so all sets will be valid).
+	filteredBestPartition := [][]*Device{}
```

Comment:
Just for my understanding, if required is not nil, should only contain one single set?

Reply:
Yes. That is correct. I would only expect one set to actually contain all of the required devices if it is not `nil`. This is because the set is part of a partition of all the devices on the machine. As such, (by definition) only one of them could ever have all of the required devices.